### PR TITLE
Zeek shaper script

### LINKED ISSUE #2104: Migrate zeek typed json to Z
(This is the Zeek equivalent of #1871 .)

### PR DESCRIPTION
**Pull request by henridf**
Also added a helper cmd `zsontypes` to help people transition away from legacy types.json. We can remove that as the same time as we remove types.json support from the ndjson reader.

close #2104

### CODE REVIEW

File: cmd/zsontypes/zsontypes.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/mccanne/charm"
+)
+
+var ZsonTypes = &charm.Spec{
+	Name:  "zsontypes",
+	Usage: "zsontypes <file.json>",
+	Short: "tool for printing zson types for legacy types.json",
+	Long: `
+The Zsontypes tool prints a ZSON-formatted for each type defined in a legacy types.json file.
+It is intended to help those transitioning from teh "types.json" approach to the newer Z-based approach.
```
Comment (philrz) [inline cmd/zsontypes/zsontypes.go:21]:
```suggestion
It is intended to help those transitioning from the "types.json" approach to the newer Z-based approach.
```

**Comment by henridf** [comment]
Some quick and dirty `time` runs show that this is significantly (over 2x) slower than types.json. A more apples-to-apples comparison is vs reading the file without applying the shaper script, and here the delta is smaller.

Applying legacy types.json:
```
$ time zq -j zeek/types.json -Z zq-sample-data/zeek-ndjson/conn.ndjson.gz > /dev/null
real	0m18.193s
user	0m40.063s
sys	0m1.480s
```

Reading the file as ndjson (no types.json) and zson):
```
$ time zq -i ndjson -Z zq-sample-data/zeek-ndjson/conn.ndjson.gz > /dev/null
real	0m52.414s
user	1m47.371s
sys	0m4.609s

$ time zq -i zson -Z zq-sample-data/zeek-ndjson/conn.ndjson.gz > /dev/null
real	0m39.450s
user	1m13.859s
sys	0m2.630s
```


Reading the file as ndjson and zson, then applying shaper.zql:

```
$ time zq -i ndjson -Z -I zeek/shaper.zql zq-sample-data/zeek-ndjson/conn.ndjson.gz > /dev/null
real	0m56.269s
user	2m4.924s
sys	0m4.827s

$ time zq -i zson -Z -I zeek/shaper.zql zq-sample-data/zeek-ndjson/conn.ndjson.gz > /dev/null
real	0m49.281s
user	1m33.752s
sys	0m2.993s
```

File: zeek/shaper.zql [hunk 2]
```diff
@@ -0,0 +1,110 @@
+type zenum=string;
+type port=uint16;
+type id={orig_h:ip,orig_p:port,resp_h:ip,resp_p:port};
```
Comment (nwt) [inline zeek/shaper.zql:3]: Nit: Sort me (and the entries in other two sections).

File: cmd/zsontypes/zsontypes.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/mccanne/charm"
+)
+
+var ZsonTypes = &charm.Spec{
+	Name:  "zsontypes",
+	Usage: "zsontypes <file.json>",
+	Short: "tool for printing zson types for legacy types.json",
+	Long: `
+The Zsontypes tool prints a ZSON-formatted for each type defined in a legacy types.json file.
```
Comment (nwt) [inline cmd/zsontypes/zsontypes.go:20]: Need a noun after "ZSON-formatted".

File: cmd/zsontypes/zsontypes.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+func init() {
+	ZsonTypes.Add(charm.Help)
+}
+
+type Command struct{}
+
+func New(f *flag.FlagSet) (charm.Command, error) {
+	c := &Command{}
+	return c, nil
```
Comment (nwt) [inline cmd/zsontypes/zsontypes.go:36]:
```suggestion
	return &Command{}, nil
```

File: cmd/zsontypes/zsontypes.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	// aliases according to its configuration.  See issue #1427.
+	_, err = zctx.LookupTypeAlias("zenum", zng.TypeString)
+	if err != nil {
+		return err
+	}
+	_, err = zctx.LookupTypeAlias("port", zng.TypeUint16)
+	if err != nil {
+		return err
+	}
+
```
Comment (nwt) [inline cmd/zsontypes/zsontypes.go:69]:
```suggestion
    // (from ndjson.io.NewReader)
    // Note: we add hardwired aliases for "port" to "uint16" when reading
    // *any* json file but they are only used when the schema mapper
    // (aka typings config) references such types from a configured schema.
    // However, the schema mapper should be responsible for creating these
    // aliases according to its configuration.  See issue #1427.
    if _, err := zctx.LookupTypeAlias("zenum", zng.TypeString); err != nil {
        return err
    }
    if _, err = zctx.LookupTypeAlias("port", zng.TypeUint16); err != nil {
        return err
    }

    tp := tzngio.NewTypeParser(zctx)
```

**Review by nwt — APPROVED** [review]

File: cmd/zsontypes/zsontypes.go (hunk 3 above)
Comment (henridf) [reply cmd/zsontypes/zsontypes.go:20]: 👍

File: zeek/shaper.zql [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+type syslog={_path:string,ts:time,uid:bstring,id:id,proto:zenum,facility:bstring,severity:bstring,message:bstring,_write_ts:time};
+type tunnel={_path:string,ts:time,uid:bstring,id:id,tunnel_type:zenum,action:zenum,_write_ts:time};
+type weird={_path:string,ts:time,uid:bstring,id:id,name:bstring,addl:bstring,notice:bool,peer:bstring,source:bstring,_write_ts:time};
+type x509={_path:string,ts:time,id:bstring,certificate:{version:uint64,serial:bstring,subject:bstring,issuer:bstring,not_valid_before:time,not_valid_after:time,key_alg:bstring,sig_alg:bstring,key_type:bstring,key_length:uint64,exponent:bstring,curve:bstring},san:{dns:[bstring],uri:[bstring],email:[bstring],ip:[ip]},basic_constraints:{ca:bool,path_len:uint64},_write_ts:time};
+
+
+put ts = iso(ts), _write_ts=iso(_write_ts) | put .=unflatten(.)
+   | switch (
+    case _path=files => put . = shape(files)
```
Comment (nwt) [inline zeek/shaper.zql:59]: Order cases by expected frequency for better performance?

**Review by nwt — APPROVED** [review]

File: zeek/shaper.zql (hunk 6 above)
Comment (nwt) [reply zeek/shaper.zql:59]: `zq -f table 'count() by _path | sort -r' zq-sample-data/zng/*` !

Comment (henridf) [reply zeek/shaper.zql:59]: Performance with/without a switch doesn't vary significantly (at least in my quick and dirty tests, where I compared with just doing `shape(conn)` on a conn log), so I'll leave this optimization for when it does.

**Pull request merged**